[now dev] Build env not passed to client

Using most up to date canary branch: 15.0.8-canary.1.

A couple of versions ago all of this was working, but now the build env variables that I'm passing to client side env are being set with `undefined`.

I have this next.config.js

```
const { PHASE_PRODUCTION_SERVER } =
  process.env.NODE_ENV === 'development'
    ? {}
    : !process.env.NOW_REGION
      ? require('next/constants')
      : require('next-server/constants');

const config = {
    target: 'serverless',
    env: {
        HASURA_URL: process.env.HASURA_URL,
        BASE_URL: process.env.BASE_URL
    }
}
module.exports = (phase, { defaultConfig }) => {
  if (phase === PHASE_PRODUCTION_SERVER) {
    // Config used to run in production.
    console.log("HASURA URL!!!", process.env.HASURA_URL)
    return config
  }
  const withCSS = require('@zeit/next-css')
  return withCSS(config);
}
```

And these are env sections from now.json

```
    "env": {
        "HASURA_URL": "http://localhost:8080/v1alpha1/graphql",
        "BASE_URL": "http://localhost:3000"
    },
    "build": {
        "env": {
            "BASE_URL": "http://localhost:3000",
            "HASURA_URL": "http://localhost:8080/v1alpha1/graphql"
        }
    },
```

Strangely this was all working perfectly, and then I added the BASE_URL value to build.env and did `now dev` in the terminal and received this error: `Cannot find module '/home/agstover/.cache/co.zeit.now/dev/builders/node_modules/@now/next'`

I ran it again and it worked, but these issues started appearing. Can't get env variables on client and a constant stream of "http://localhost:3000/_next/on-demand-entries-ping?page=/ 404 (Not Found)" in the browser console.

The repo can be found here: https://github.com/agstover/awesomex

--- a/src/commands/dev/lib/dev-server.ts
+++ b/src/commands/dev/lib/dev-server.ts
@@ -295,7 +295,7 @@ export default class DevServer {
     }
   }
 
-  async getLocalEnv(fileName: string): Promise<EnvConfig> {
+  async getLocalEnv(fileName: string, base?: EnvConfig): Promise<EnvConfig> {
     // TODO: use the file watcher to only invalidate the env `dotfile`
     // once a change to the `fileName` occurs
     const filePath = join(this.cwd, fileName);
@@ -309,7 +309,7 @@ export default class DevServer {
         throw err;
       }
     }
-    return env;
+    return { ...base, ...env };
   }
 
   async getNowJson(): Promise<NowConfig> {
@@ -398,13 +398,15 @@ export default class DevServer {
 
     // Retrieve the path of the native module
     const modulePath = await getModuleForNSFW(this.output);
+    const nowJson = await this.getNowJson();
+    const nowJsonBuild = nowJson.build || {};
     const [env, buildEnv] = await Promise.all([
-      this.getLocalEnv('.env'),
-      this.getLocalEnv('.env.build')
+      this.getLocalEnv('.env', nowJson.env),
+      this.getLocalEnv('.env.build', nowJsonBuild.env)
     ]);
+    Object.assign(process.env, buildEnv);
     this.env = env;
     this.buildEnv = buildEnv;
-    const nowJson = await this.getNowJson();
 
     const opts = { output: this.output, isBuilds: true };
     const files = await getFiles(this.cwd, nowJson, opts);

--- a/src/commands/dev/lib/dev-server.ts
+++ b/src/commands/dev/lib/dev-server.ts
@@ -125,9 +125,7 @@ export default class DevServer {
 
     // Update the build matches in case an entrypoint was created or deleted
     const nowJson = await this.getNowJson();
-    if (nowJson) {
-      await this.updateBuildMatches(nowJson);
-    }
+    await this.updateBuildMatches(nowJson);
 
     const filesChangedArray = [...filesChanged];
     const filesRemovedArray = [...filesRemoved];
@@ -309,6 +307,7 @@ export default class DevServer {
         throw err;
       }
     }
+    this.validateEnvConfig(fileName, base || {}, env);
     return { ...base, ...env };
   }
 
@@ -356,12 +355,6 @@ export default class DevServer {
     if (config.version !== 2) {
       throw new Error('Only `version: 2` is supported by `now dev`');
     }
-    this.validateEnvConfig('.env', config.env, this.env);
-    this.validateEnvConfig(
-      '.env.build',
-      config.build && config.build.env,
-      this.buildEnv
-    );
   }
 
   validateEnvConfig(

